cache/fifo_test.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package cache
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestFIFOCache(t *testing.T) {
+	tests := []struct {
+		name     string
+		limit    int
+		maxVal   int
+		expected error
+		check    func(t *testing.T, cache *FIFO[int, int])
+	}{
+		{
+			name:     "empty cache fails",
+			limit:    0,
+			expected: errors.New("maxSize must be greater than 0"),
+		},
+		{
+			name:   "inserting up to limit works",
+			limit:  10,
+			maxVal: 9,
+		},
+		{
+			name:   "inserting after limit cleans first",
+			limit:  10,
+			maxVal: 10,
+			check: func(t *testing.T, cache *FIFO[int, int]) {
+				require := require.New(t)
+				_, ok := cache.Get(0)
+				require.False(ok)
+				v, ok := cache.Get(1)
+				require.True(ok)
+				require.Equal(1, v)
+			},
```

Comment:
I think these tests are on the right track but still find that these are a little hard to reason about (particularly the limit/maxval) parameter are not obvious on how they work until you read the test body.

I think one way we could make this cleaner (and remove the `check` field) would be to make a struct like so:

```
func TestFIFOCache(t *testing.T) {
...
type put struct {
    i     int
    ok  bool
}

type get struct {
    i     int
    ok  bool
}
```

and if we did this I think we could write tests as:
```
{
    puts:[]put{
        {i: 1, ok: true}
        {i: 1, ok: false}
    }
    gets:[]get{
        {i: 1,	ok: true}
    }
}
```

Reply:
This structure looks like a nice improvement. This should make it a bit more readable and we can just use a size like 2, so it's not excessive.

I think we should test the following cases here:
- no elements removed when the cache is less than full
- no elements removed when cache is exactly full
- elements removed in FIFO order when cache overfills
- elements removed in FIFO order after additional get/put ops for element to be removed FIFO

I think we're currently missing the last case. There's a unit test to make sure if we insert the first element again, then it stays in the cache, but we should additionally test that this does not change its position in the queue for removal to differentiate FIFO from LRU.

Reply:
Thanks both !
I have changed the test suite to only have a `name`, `osp` pairs and it's way more compact/readable.
Also added tests cases to cover @aaronbuchwald ones, haven't implemented the last one for now (TODO!)

Reply:
I'm not sure if there is a real difference between the two last test cases, the case "elements removed in FIFO order when cache overfills" seems to fulfil both, please tell me if that's wrong.

Reply:
Actually they are different, see the test case "elements removed in FIFO order and not LRU".